src/d3/hapticplot.d3.ts

```diff
@@ -0,0 +1,137 @@
+// run `tsc Hapticplot.ts` to compile into Hapticplot.js file
+
+// for running on browser
+// import * as _d3 from 'd3';
+
+// declare global {
+// const d3: typeof _d3;
+// }
+
+// for running on unit tests
+import * as d3 from 'd3';
+import * as THREE from 'three';
+
+export class Hapticplot{
+    private data: number[];
+    private shape: string;
+    private container: HTMLElement | null;
+    private hscale: d3.ScaleLinear<number, number>;
+
+  constructor(shape: string) {
+    this.shape = shape;
+  }
+  init(container: HTMLElement | null, data: number[]){
+    this.data = data;
+    this.container = container;
+    // create a scale so that there is correspondence between data set and screen render
+    this.hscale = d3.scaleLinear();
+    this.hscale.domain([0, d3.max(this.data) as number])       // max of dataset
+    .range([0, 100]);                                      // linear mapping of data set values to values from 0 to 10
+    this.createSky();
+    this.generatePts();
+    this.setColor('blue');
+    this.setRadius(0.1);
+    this.setInteraction();
+    this.addEventListeners();
+    this.createGridPlane();
```

Comment:
Rather than having a method for each piece of the d3 initialization, it's standard to just do it all in one cascade, since they rarely need to be updated independently. That way when the data updates later, you don't have to go through and call all the methods again.

What I mean by one big cascade is like
```typescript
d3.select(this.container).selectAll(this.shape)
  .attr('position', position)
  .attr('radius', radius)
  .attr('color', color);
```
And so on.

If you do end up doing that, then you typically wouldn't put things like callback functions inline; you'd define them elsewhere.

Reply:
I condensed them all into one cascading function! I'm having a bit of trouble defining the callbacks outside of it though. For the first call back, (which sets the objects positions) I need access to the (d, i) vales from the d3 selection, but I was able to pull the logic out into a helper function called "generatePositions." For the event listeners callback, however, I need to maintain access to the event and the currently selected entity, which I'm not sure how to maintain when passing it to the helper.